ramips: add support for Xiaomi Mi Router CR660x series

IgnotaYun opened the pull request:
Xiaomi Mi Router CR6606 is a Wi-Fi6 AX1800 Router with four GbE Ports.
Alongside the general model, it has three carrier customized models:
CR6606 (China Unicom), CR6608 (China Mobile), CR6609 (China Telecom)

Specifications:
- SoC: MediaTek MT7621AT
- RAM: 256MB DDR3 (ESMT M15T2G16128A)
- Flash: 128MB NAND (ESMT F59L1G81MB)
- Ethernet: 1000Base-T x4 (MT7530 SoC)
- WLAN: 2x2 2.4GHz 574Mbps + 2x2 5GHz 1201Mbps (MT7905DAN + MT7975DN)
- LEDs: System (Blue, Yellow), Internet (Blue, Yellow)
- Buttons: Reset, WPS
- UART: through-hole on PCB (\[VCC 3.3v\]\(RX\)(GND)(TX) 115200, 8n1)
- Power: 12VDC, 1A

Jailbreak Notes:
1. Get shell access.
   (See the commit message for details)
2. There's no official binary firmware available, and if you lose the content of your flash, no one except Xiaomi can help you. Dump these partitions in case you need them: "Bootloader" "Nvram" "Bdata" "crash" "crash_log" "firmware" "firmware1" "overlay" "obr"
   (See the commit message for details)
3. Since UART access is locked ootb, you should get permanent UART access by editing uboot env. Otherwise, your router may become bricked.
   (See the commit message for details)

Migrate to OpenWrt:
1. Transfer squashfs-firmware.bin to the router. 
2. nvram set flag_try_sys1_failed=0
3. nvram set flag_try_sys2_failed=1
4. nvram commit
5. mtd -r write /path/to/image/squashfs-firmware.bin firmware

Additional Info:
1. CR660x series routers have a different nand layout compared to other Xiaomi nand devices.
2. This router has a relatively fresh uboot (2018.09) compared to other Xiaomi devices, and it is capable of booting fit image firmware. Unfortunately, no successful attempt of booting OpenWrt fit image were made so far. The cause is still yet to be known. For now, we use legacy image instead.

Signed-off-by: Raymond Wang <infiwang@pm.me>

infiWang commented:
By the way, this patch could be backported to 21.02 seamlessly.

hauke commented on target/linux/ramips/dts/mt7621_xiaomi_mi-router-cr6606.dts:
```diff
@@ -0,0 +1,187 @@
+		partition@7000000 {
+			label = "obr";
+			reg = <0x7000000 0x1000000>;
+            read-only;
```
The indention looks wrong here

infiWang replied:
Fixed.

namidairo commented:
> Get shell access using the Redmi AX6 method.
 Ref: https://forum.openwrt.org/t/ssh-router-redmi-ax6/4

I think you've truncated that link there, looks like it 404's.

Presumably you meant https://forum.openwrt.org/t/ssh-router-redmi-ax6/82893/4

I'm not sure linking to a forum post is the best way to document this.

> Since uart access is locked ootb, you should get permanent uart access by unlocking and editing Bdata. Otherwise, you may get your device bricked and it's difficult to repair.
 Ref: https://forum.openwrt.org/t/xiaomi-ax3600-ssh-guide/65438/

Seems to be talking about ssh only in that post. Presumably you'd need to modify these requests to set `uart_en=1` and `boot_wait=on` to achieve working uart input and the ability to interrupt uboot. I think there may have been a third variable for the seconds as well that might have needed to be set as well, if I remember correctly. Document these formally.

I also feel like we could set one of the boot failure flags to force it to boot from firmware1 and then fwconcat with the overlay for extra space, but that sounds like it could be more trouble than it's worth for an extra 30mb of space.

infiWang commented:
> > Get shell access using the Redmi AX6 method.
> > Ref: https://forum.openwrt.org/t/ssh-router-redmi-ax6/4
> 
> I think you've truncated that link there, looks like it 404's.
> 
> Presumably you meant https://forum.openwrt.org/t/ssh-router-redmi-ax6/82893/4
> 
> I'm not sure linking to a forum post is the best way to document this.
> 
> > Since uart access is locked ootb, you should get permanent uart access by unlocking and editing Bdata. Otherwise, you may get your device bricked and it's difficult to repair.
> > Ref: https://forum.openwrt.org/t/xiaomi-ax3600-ssh-guide/65438/
> 
> Seems to be talking about ssh only in that post. Presumably you'd need to modify these requests to set `uart_en=1` and `boot_wait=on` to achieve working uart input and the ability to interrupt uboot. I think there may have been a third variable for the seconds as well that might have needed to be set as well, if I remember correctly. Document these formally.

Thanks for your advice, I've added details to the commit message. The problem for me is that it might be too long.

In the meantime, I believe we should have a device page on the OpenWrt Wiki.

> I also feel like we could set one of the boot failure flags to force it to boot from firmware1 and then fwconcat with the overlay for extra space, but that sounds like it could be more trouble than it's worth for an extra 30mb of space.

1.  From what I know, this is possible but would make revert from OpenWrt to stock painful. This model has no official firmware binary and is not supported by the recovery tool according to Xiaomi. I tried to keep the partition table mostly the same for this reason. 
2. Other Xiaomi nand devices which have the same nand model suffer from occasional bad block problems, while a layout like stock firmware do mitigate most of them from my observation. However, my testing CR6606, CR6608, and CR6609 works like a charm and we do have to admit that most routers out there don't have similar issues at all. Maybe we should add an extra "flavor"?

infiWang commented:
I believe I've found another way to unlock the UART lock, please give me a day or two to update the commit message.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1387,6 +1387,33 @@ define Device/xiaomi_mi-router-ac2100
+  DEVICE_ALT1_VARIANT := China Mobile
+  DEVICE_ALT2_VENDOR := Xiaomi
+  DEVICE_ALT2_MODEL := Mi Router CR6609
+  DEVICE_ALT2_VARIANT := China Telecom
```
Consider dropping the *VARIANT variables here, as they don't look like they are used "correctly" (i.e. they don't really  help here at the moment).

infiWang replied:
Basically, CR6608 and CR6609 both only have carrier customized variants for the time being. I'm curious whether Xiaomi would release a general variant like CR6606, so I tried to use it as an addition.
After some thought, I believe it would be nonsense in the build system. Dropped all the variants for now.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1387,6 +1387,33 @@ define Device/xiaomi_mi-router-ac2100
+  IMAGE_SIZE := 51200k
+  IMAGES += firmware.bin overlay.bin
+  IMAGE/firmware.bin := append-kernel | pad-to 128k | append-rootfs | \
+  pad-rootfs
```
one tab for hanging indent.

infiWang commented:
Got some schoolwork to do lately.
It should be better now.

adschm commented on target/linux/ramips/dts/mt7621_xiaomi_mi-router-cr6606.dts:
```diff
@@ -0,0 +1,187 @@
+		};
+	};
+
+	keys: keys {
```
The "keys:" DT label is not used an can be removed.

981213 commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1394,6 +1394,28 @@ define Device/xiaomi_mi-router-ac2100
 
+define Device/xiaomi_mi-router-cr6606
+  $(Device/dsa-migration)
+  $(Device/uimage-lzma-loader)
```
The lzma loader is for the ancient u-boot 1.1.3 which can't properly extract a big kernel. The new u-boot 2018 from mediatek has proper lzma decompression support and lzma-loader isn't needed. This line can be dropped.

infiWang replied:
Give a shot, got stuck like a FIT image while booting. Not sure why.

981213 replied:
You can't change your load address in mt7621.mk without a loader that can relocate itself (e.g. we have lzma-loader and relocate-kernel). This load address must match what the kernel makefile says when using fit or legacy lzma uimage.
Continue using lzma loader here then.

infiWang replied:
Thanks, no idea about the build behavior before.
Tried relocate-kernel with both fit image and legacy image, still can't make it boot. The uboot reversion is 2018.09. Stick with lzma-loader for now.

981213 commented on target/linux/ramips/dts/mt7621_xiaomi_mi-router-cr6606.dts:
```diff
@@ -0,0 +1,187 @@
+			read-only;
+		};
+
+		partition@200000 {
```
This should be a kernel partition, not a partition for the entire firmware. The label should be 'kernel' and the compatible string should be removed.

infiWang replied:
Actually, CR660x series factory firmware layout looks exactly the same. For OpenWRT though, it do worth a try.
By the way, is the "kernel-ubi" layout good for FIT image? The current legacy image is a result of an unbootable OpenWrt FIT image, maybe it will change later one day.

infiWang replied:
Overhauled the partition table, squashed most partitions into one "firmware". No recovery tool available so I don't see the need of keeping a stock "firmware", and 30MB for kernel seems like a waste.

AmadeusGhost replied:
However, it doesn't seem right here. Generally, `compatible = "denx, uimage"` is not used on nand devices.
```
define Device/xiaomi_mi-router-cr6606
  ...
  KERNEL_SIZE := 4096k
  IMAGE_SIZE := 128512k
```
Since the size of the kernel partition is limited here, the partition table of dts may should be changed to this:
```
partition@200000 {
	label = "kernel";
	reg = <0x200000 0x400000>;
};

partition@600000 {
	label = "ubi";
	reg = <0x600000 0x7980000>;
};
```

infiWang replied:
I see. Take Netis WF2881's approach seems to be a solution.

infiWang replied:
Fixed.

981213 commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1394,6 +1394,28 @@ define Device/xiaomi_mi-router-ac2100
+  IMAGE_SIZE := 51200k
+  IMAGES += firmware.bin overlay.bin
+  IMAGE/firmware.bin := append-kernel | pad-to 128k | append-rootfs | \
+	pad-rootfs
```
Everything after append-kernel should be removed. You only need the kernel part here. rootfs should be mounted from ubi and this appended rootfs is useless.

infiWang replied:
Thanks, I'll take a look very soon.

981213 reviewed (changes requested):
Hi! This partition table looks wrong. For a nand device, rootfs is included in ubi and we only put a kernel in the original firmware partition for u-boot to load. With your current setup, a kernel partition with the exact size of your kernel image will be split from firmware and sysupgrade will fail if the kernel grows in size in the future.
Please test if the following changes work:

stevenliuit commented:
why？

infiWang commented:
> why？

Why what? Those requests do get their point, and I've finished these changes.

stevenliuit commented:
> > why？
> 
> Why what? Those requests do get their point, and I've finished these changes.

I think this request is meaningful, why was it rejected?

infiWang commented:
> > > why？
> > 
> > Why what? Those requests do get their point, and I've finished these changes.
> 
> I think this request is meaningful, why was it rejected?

Presumably you mean "this PR"? It's still ongoing due to some former issues.

infiWang commented:
@adschm @981213
Sorry for disturbing you. I wonder whether any further changes are needed here, or any extra setup I should take?

infiWang commented:
@981213 Hi, is it convenient for you to review my changes?

1715173329 commented:
Just my personal view: since these boards share the same hardware and the only difference among them is the naming, why not directly merge them into a single profile like "cr660x"?
Thanks for your hard work.

adschm commented:
> Just my personal view: since these boards share the same hardware and the only difference among them is the naming, why not directly merge them into a single profile like "cr660x"?
> Thanks for your hard work.

Can you say 100% that no other device matching this pattern will come up? So far, it's only 3/10.

Are you sure they are 100% equal, so nobody would require to use the name for filtering some behavior in the future?

1715173329 commented:
> Can you say 100% that no other device matching this pattern will come up? So far, it's only 3/10.

In general there won't come up new board named cr660x (according to Xiaomi's naming convention), however I'm not their staff so obviously I cannot promise that.

> 
> Are you sure they are 100% equal, so nobody would require to use the name for filtering some behavior in the future?

These 3 boards are 100% equal at least, the different names are just for the different ISPs mentioned in PR message.

infiWang commented:
> These 3 boards are 100% equal at least, the different names are just for the different ISPs mentioned in PR message.

Not quite in my opinion: every model has a slightly different bootloader, some envs are model specific too.

Still, I do have to admit that they are binary compatible. Since there was no such precedent in-tree, my decision was to follow "Linksys" devices' choice and leave it for reviewers to judge.

IgnotaYun committed:
ramips: add support for Xiaomi Mi Router CR660x series

Xiaomi Mi Router CR6606 is a Wi-Fi6 AX1800 Router with 4 GbE Ports.
Alongside the general model, it has three carrier customized models:
CR6606 (China Unicom), CR6608 (China Mobile), CR6609 (China Telecom)

Specifications:
- SoC: MediaTek MT7621AT
- RAM: 256MB DDR3 (ESMT M15T2G16128A)
- Flash: 128MB NAND (ESMT F59L1G81MB)
- Ethernet: 1000Base-T x4 (MT7530 SoC)
- WLAN: 2x2 2.4GHz 574Mbps + 2x2 5GHz 1201Mbps (MT7905DAN + MT7975DN)
- LEDs: System (Blue, Yellow), Internet (Blue, Yellow)
- Buttons: Reset, WPS
- UART: through-hole on PCB ([VCC 3.3v](RX)(GND)(TX) 115200, 8n1)
- Power: 12VDC, 1A

Jailbreak Notes:
1. Get shell access.
   1.1. Get yourself a wireless router that runs OpenWrt already.
   1.2. On the OpenWrt router:
      1.2.1. Access its console.
      1.2.2. Create and edit
             /usr/lib/lua/luci/controller/admin/xqsystem.lua
             with the following code (exclude backquotes and line no.):
```
     1  module("luci.controller.admin.xqsystem", package.seeall)
     2
     3  function index()
     4      local page   = node("api")
     5      page.target  = firstchild()
     6      page.title   = ("")
     7      page.order   = 100
     8      page.index = true
     9      page   = node("api","xqsystem")
    10      page.target  = firstchild()
    11      page.title   = ("")
    12      page.order   = 100
    13      page.index = true
    14      entry({"api", "xqsystem", "token"}, call("getToken"), (""),
103, 0x08)
    15  end
    16
    17  local LuciHttp = require("luci.http")
    18
    19  function getToken()
    20      local result = {}
    21      result["code"] = 0
    22      result["token"] = "; nvram set ssh_en=1; nvram commit; sed -i
's/channel=.*/channel=\"debug\"/g' /etc/init.d/dropbear; /etc/init.d/drop
bear start;"
    23      LuciHttp.write_json(result)
    24  end
```
      1.2.3. Browse http://{OWRT_ADDR}/cgi-bin/luci/api/xqsystem/token
             It should give you a respond like this:
             {"code":0,"token":"; nvram set ssh_en=1; nvram commit; ..."}
             If so, continue; Otherwise, check the file, reboot the rout-
             er, try again.
      1.2.4. Set wireless network interface's IP to 169.254.31.1, turn
             off DHCP of wireless interface's zone.
      1.2.5. Connect to the router wirelessly, manually set your access
             device's IP to 169.254.31.3, make sure
             http://169.254.31.1/cgi-bin/luci/api/xqsystem/token
             still have a similar result as 1.2.3 shows.
   1.3. On the Xiaomi CR660x:
        1.3.1. Login to the web interface. Your would be directed to a
               page with URL like this:
               http://{ROUTER_ADDR}/cgi-bin/luci/;stok={STOK}/web/home#r-
               outer
        1.3.2. Browse this URL with {STOK} from 1.3.1, {WIFI_NAME}
               {PASSWORD} be your OpenWrt router's SSID and password:
               http://{MIROUTER_ADDR}/cgi-bin/luci/;stok={STOK}/api/misy-
               stem/extendwifi_connect?ssid={WIFI_NAME}&password={PASSWO-
               RD}
               It should return 0.
        1.3.3. Browse this URL with {STOK} from 1.3.1:
               http://{MIROUTER_ADDR}/cgi-bin/luci/;stok={STOK}/api/xqsy-
               stem/oneclick_get_remote_token?username=xxx&password=xxx&-
               nonce=xxx
   1.4. Before rebooting, you can now access your CR660x via SSH.
        For CR6606, you can calculate your root password by this project:
        https://github.com/wfjsw/xiaoqiang-root-password, or at
        https://www.oxygen7.cn/miwifi.
        The root password for carrier-specific models should be the admi-
        nistration password or the default login password on the label.
        It is also feasible to change the root password at the same time
        by modifying the script from step 1.2.2.
        You can treat OpenWrt Router however you like from this point as
        long as you don't mind go through this again if you have to expl-
        oit it again. If you do have to and left your OpenWrt router unt-
        ouched, start from 1.3.
2. There's no official binary firmware available, and if you lose the
   content of your flash, no one except Xiaomi can help you.
   Dump these partitions in case you need them:
   "Bootloader" "Nvram" "Bdata" "crash" "crash_log"
   "firmware" "firmware1" "overlay" "obr"
   Find the corespond block device from /proc/mtd
   Read from read-only block device to avoid misoperation.
   It's recommended to use /tmp/syslogbackup/ as destination, since files
   would be available at http://{ROUTER_ADDR}/backup/log/YOUR_DUMP
   Keep an eye on memory usage though.
3. Since UART access is locked ootb, you should get UART access by modify
   uboot env. Otherwise, your router may become bricked.
   Excute these in stock firmware shell:
    a. nvram set boot_wait=on
    b. nvram set bootdelay=3
    c. nvram commit
   Or in OpenWrt:
    a. opkg update && opkg install kmod-mtd-rw
    b. insmod mtd-rw i_want_a_brick=1
    c. fw_setenv boot_wait on
    d. fw_setenv bootdelay 3
    e. rmmod mtd-rw

Migrate to OpenWrt:
 1. Transfer squashfs-firmware.bin to the router.
 2. nvram set flag_try_sys1_failed=0
 3. nvram set flag_try_sys2_failed=1
 4. nvram commit
 5. mtd -r write /path/to/image/squashfs-firmware.bin firmware

Additional Info:
 1. CR660x series routers has a different nand layout compared to other
    Xiaomi nand devices.
 2. This router has a relatively fresh uboot (2018.09) compared to other
    Xiaomi devices, and it is capable of booting fit image firmware.
    Unfortunately, no successful attempt of booting OpenWrt fit image
    were made so far. The cause is still yet to be known. For now, we use
    legacy image instead.

Signed-off-by: Raymond Wang <infiwang@pm.me>

infiWang commented:
Migrated mtd-mac-address to nvmem implementation. Tested to be working.

DragonBluep commented:
The stock firmware is FIT image, although legacy uImage can also be accepted by u-boot. [boot log](https://www.right.com.cn/forum/thread-7485376-1-1.html)
Ramips platform currently does not have any device that uses FIT Image for reference so I think uImage is okay.

hauke commented:
Thank you for your patch, I applied it to master in 3343ca7e6837b2ac5f237ea78bf73d50831dea20.
